### PR TITLE
i382 Add more `Boundaries` types

### PR DESCRIPTION
**Pull request by MichaelClerx**
See #380 #382 

This PR:
- Splits the class `Boundaries` into an interface `Boundaries` and an implementation `RectangularBoundaries`
- Updates the optimisers so allow arbitrary boundary shapes (not just rectangular). This is a bit of extra code in each optimiser to keep track of which points we asked the user to evaluate (leaving out the out-of-bounds ones for speed)
- Adds a `LogPDFBoundaries` object that lets you use a LogPrior (with a user specified cut-off value) as boundaries for an optimisation
- Adds a ParabolicError toy error measure, to speed up unit tests for optimisers

### CODE REVIEW

**Comment by MichaelClerx** [comment]
Having more complex boundaries creates some issues for optimisers, as we can't use the triangle transform! So this is not ready to merge yet...

**Comment by MichaelClerx** [comment]
Closes #380
Closes #382

**Comment by MichaelClerx** [comment]
Ready for review now @martinjrobins :-)

File: pints/_boundaries.py [hunk 1]
```diff
@@ -13,6 +13,38 @@
...
+
+    def sample(self, n=1):
+        """
+        Returns ``n`` random samples from within the boundaries.
+
+        The returned value is a numpy array with shape ``(n, d)`` where ``n``
+        is the requested number of samples, and ``d`` is the dimension of the
+        parameter space these boundaries are defined on.
+
+        *Note that the :meth:`sample()` method does not always have to be
```
Comment (martinjrobins) [inline pints/_boundaries.py]: might want to specify the cases in which the `sample` method *does have to be implemented

**Pull request merged**